torchrl/data/rlhf/dataset.py

```diff
@@ -0,0 +1,348 @@
+        from_disk (bool, optional): if ``True``, :func:`datasets.load_from_disk`
+            will be used. Otherwise, :func:`datasets.load_dataset` will be used.
+            Defaults to ``False``.
+
+    The dataset will be stored in ``root/<split>/<max_length>/``
+    Examples:
+        >>> from torchrl.data.rlhf.comparison import make_process_fn_comparison, pre_tokenization_hook
+        >>> split = "train"
+        >>> max_length = 550
+        >>> dataset_name = "CarperAI--openai_summarize_comparisons"
+        >>> dataset = create_or_load_dataset(
+        ...     split,
+        ...     max_length,
+        ...     dataset_name,
+        ...     make_process_fn_comparison,
+        ...     pre_tokenization_hook=pre_tokenization_hook,
+        ... )
+        >>> print(dataset)
+        TensorDict(
+            fields={
+                train: TensorDict(
+                    fields={
+                        550: TensorDict(
+                            fields={
+                                attention_mask: MemmapTensor(shape=torch.Size([185068, 550]), device=cpu, dtype=torch.int64, is_shared=False),
+                                input_ids: MemmapTensor(shape=torch.Size([185068, 550]), device=cpu, dtype=torch.int64, is_shared=False)},
+                            batch_size=torch.Size([185068, 550]),
+                            device=None,
+                            is_shared=False)},
+                    batch_size=torch.Size([185068, 550]),
+                    device=None,
+                    is_shared=False)},
+            batch_size=torch.Size([185068, 550]),
+            device=None,
+            is_shared=False)
+
+    """
+    if root_dir is None:
+        root_dir = Path(os.environ.get("HOME")) / ".cache/torchrl/data/"
+        os.makedirs(root_dir, exist_ok=True)
+    root_dir = Path(root_dir)
+    data_dir = root_dir / str(Path(dataset_name).name).split("-")[0]
+    data_dir_total = data_dir / split / str(max_length)
+    # search for data
+    if os.path.exists(data_dir_total):
+        dataset = TensorDict.load_memmap(data_dir_total)
+        return dataset
+    dataset = load_dataset(
+        split=split,
+        dataset_name=dataset_name,
+        pre_tokenization_hook=pre_tokenization_hook,
+        from_disk=from_disk,
+    )
+    dataset = tokenize(
+        dataset,
+        max_length=max_length,
+        make_process_fn=make_process_fn,
+    )
+    prefix = (split, str(max_length))
+    return dataset_to_tensordict(
+        dataset, data_dir=data_dir, prefix=prefix, valid_mask_key="valid_sample"
+    )[prefix]
+
+
+def load_dataset(
+    split,
+    dataset_name,
+    pre_tokenization_hook=None,
+    from_disk: bool = False,
+):
+    """Loads a text dataset from ``datasets``.
+
+    Args:
+        split (str): One of ``"train"`` or ``"valid"``.
+        dataset_name (str): the name or path of the dataset.
+        pre_tokenization_hook (callable): TODO
```

Comment:
`pre_tokenization_hook` is called on the Dataset before tokenization and should return a modified Dataset object. The intended use is for carrying out tasks that require modifying the dataset as a whole as opposed to modifying individual datapoints, for example discarding certain datapoints based on a particular condition. Tokenization and other "elementwise" operations on the data are performed by the process function which is mapped over the dataset.